ext/datadog_profiling_native_extension/heap_recorder.c

```diff
@@ -933,16 +858,15 @@ heap_stack* heap_stack_new(ddog_prof_Slice_Location locations) {
   return stack;
 }
 
-void heap_stack_free(heap_stack *stack) {
-  for (uint64_t i = 0; i < stack->frames_len; i++) {
-    heap_frame *frame = &stack->frames[i];
-    ruby_xfree(frame->name);
-    ruby_xfree(frame->filename);
+void heap_stack_free(heap_recorder *recorder, heap_stack *stack) {
+  for (u_int16_t i = 0; i < stack->frames_len; i++) {
+    unintern_or_raise(recorder, stack->frames[i].filename);
+    unintern_or_raise(recorder, stack->frames[i].name);
   }
+
   ruby_xfree(stack);
 }
 
-// WARN: Must be kept in-sync with ::ddog_location_slice_hash
 st_index_t heap_stack_hash(heap_stack *stack, st_index_t seed) {
   st_index_t hash = seed;
   for (uint64_t i = 0; i < stack->frames_len; i++) {
```

Comment:
Now that this is all a bunch of numbers, I wonder if we could hash it all in one go (rather than going frame by frame).

Reply:
Yes this along with the cmp version are great points!